### PR TITLE
Read logs from all containers in pod for KubernetesPodOperator

### LINKED ISSUE #27282: KubernetesPodOperator: Option to show logs from all containers in a pod
### Description

Currently, KubernetesPodOperator fetches logs using
```
             self.pod_manager.fetch_container_logs(
                    pod=self.pod,
                    container_name=self.BASE_CONTAINER_NAME,
                    follow=True,
                )
```
and so only shows log from the main container in a pod. It would be very useful/helpful to have the possibility to fetch logs for all the containers in a pod.

### Use case/motivation

Making the cause of failed KubernetesPodOperator tasks a lot more visible.

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by mlnsharma**
Implemented changes to collect logs from all containers (except airflow_xcom_sidecar container) of pod, when using KubernetesPodOperator in a DAG with get_logs=True option.

### CODE REVIEW

**Comment by mlnsharma** [comment]
Hi @jedcunningham This PR fixes the issue from #27282

File: airflow/providers/cncf/kubernetes/utils/pod_manager.py [hunk 1]
```diff
@@ -234,11 +234,12 @@ def consume_logs(*, since_time: DateTime | None = None, follow: bool = True) ->
                 for raw_line in logs:
                     line = raw_line.decode("utf-8", errors="backslashreplace")
                     timestamp, message = self.parse_log_line(line)
-                    self.log.info(message)
+                    self.log.info('[%s] %s', container_name, message)
```
Comment (bdsoha) [inline airflow/providers/cncf/kubernetes/utils/pod_manager.py]: Use double quotes:

```suggestion
                    self.log.info("[%s] %s", container_name, message)
```

File: airflow/providers/cncf/kubernetes/utils/pod_manager.py [hunk 2]
```diff
@@ -345,6 +356,17 @@ def read_pod_logs(
...
             self.log.exception("There was an error reading the kubernetes API.")
             raise
+    
+    @tenacity.retry(stop=tenacity.stop_after_attempt(3), wait=tenacity.wait_exponential(), reraise=True)
+    def get_container_names(self, pod: V1Pod):
+        """Read Container names from POD"""
+        container_names = []
+        pod_info = self.read_pod(pod)
+        for container_spec in pod_info.spec.containers:
+            if(container_spec.name != 'airflow-xcom-sidecar'):
```
Comment (bdsoha) [inline airflow/providers/cncf/kubernetes/utils/pod_manager.py]:
```suggestion
            if(container_spec.name != "airflow-xcom-sidecar"):
```

**Review by bdsoha — CHANGES_REQUESTED** [review]

**Comment by bdsoha** [comment]
@mlnsharma It seems like you have failing tests related to k8s.

https://github.com/apache/airflow/actions/runs/3507624185/jobs/5875524186#step:7:463

**Review by XD-DENG — COMMENTED** [review]
This change changes the behavior without "backward compatibility".

For example, what if I as a user still prefer to read log ONLY from he base container?

**Comment by mlnsharma** [comment]
Thanks @bdsoha @XD-DENG @pulquero for your comments.
I will fix the double quotes, test failures. Regarding the backward compatibility. Do you all agree with the below solution:
- 'log_containers' will be a new parameter in KubernetesPodOperator of type string array, with the default value being ['base']. This will work only if 'get_logs=True' is passed. If the 'log_containers' value is passed as any empty array [], then all containers excluding airflow-xcom-sidecar container logs will be displayed. If the value contains any other set of container names, all those container logs will only be captured.  This new param will also be added to documentation.
- Since Airflow defaults the name of the first container to 'base', if the 'log_containers' param value consists of the actual name provided in the 'pod_template_file', all the 'log_containers' names will be checked against the actual container names in the pod. If a container name is not present in the actual pod, it will be treated as the 'base' container.

Please let me know your thoughts on the param name and any of the above statements.

**Comment by bdsoha** [comment]
> Thanks @bdsoha @XD-DENG @pulquero for your comments. I will fix the double quotes, test failures. Regarding the backward compatibility. Do you all agree with the below solution:
> 
> * 'log_containers' will be a new parameter in KubernetesPodOperator of type string array, with the default value being ['base']. This will work only if 'get_logs=True' is passed. If the 'log_containers' value is passed as any empty array [], then all containers excluding airflow-xcom-sidecar container logs will be displayed. If the value contains any other set of container names, all those container logs will only be captured.  This new param will also be added to documentation.
> * Since Airflow defaults the name of the first container to 'base', if the 'log_containers' param value consists of the actual name provided in the 'pod_template_file', all the 'log_containers' names will be checked against the actual container names in the pod. If a container name is not present in the actual pod, it will be treated as the 'base' container.
> 
> Please let me know your thoughts on the param name and any of the above statements.

The proposed `log_containers` argument should allow for a `list[str] | bool`, when the value is set to `True` all containers will be logged. This simplifies the API, IMO.

**Comment by potiuk** [comment]
Errors in tests.

**Comment by mlnsharma** [comment]
Hi @bdsoha @potiuk I fixed the code as per comments, implemented a new log_containers option, added a couple of unit tests for the same. 
When I was testing with 'get_logs=False' option, I noticed that the airflow-xcom-sidecar container was not getting started and pod changed to NotReady state. I fixed the await_container_completion code and then it started working, I included that change also in the PR.
Can you please run the tests workflow? and review my commits?

PS: I am not sure how to squash all my commits in the PR.

**Comment by potiuk** [comment]
> PS: I am not sure how to squash all my commits in the PR.

For example you can use GitHubDesktop - but there are many other guides in the internet you can find:

https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/managing-commits/squashing-commits

**Comment by XD-DENG** [comment]
Eventually when we merge in the GitHub UI, we can choose "Squash and Merge" option. That should be sufficient?

**Comment by potiuk** [comment]
> Eventually when we merge in the GitHub UI, we can choose "Squash and Merge" option. That should be sufficient?

When we merge yes. But squashing is really helpful when you continue rebasing stuff while new commits are coming on main. The problem when you `merge` using the UI rather than rebase is that your PR history gets like this:

Screenshot 2022-12-03 at 18 30 39

Not a problem just yet in this case - but with a number of merge commits for the past, this leads to the problem that you cannot easily rebase any more because those past merges are often conflicting with current main.

So  in our case where you keep on rebasing, that becomes more and more problematic (and then eventually we cannot help by rebasing the PR while reviewing it using "Update with rebase". 

Screenshot 2022-12-03 at 18 35 33

This is often useful, when we fix things in main - we can rebase a PR without asking a contributor to do so. This is  for exmple #26970 - I have to always ask the author to do it now.

On the other hand, when you squash those changes (including past merges)  and rebase, further rebasing becomes painless - because you effectively remove all those merges and what remains is only your change - solving conflicts on rebasing then is much simpler.

**Comment by mlnsharma** [comment]
Hi @potiuk @XD-DENG @bdsoha Thanks for your comments. I squashed all the commits into one commit now. Can you please review it again?

**Comment by mlnsharma** [comment]
Hi @potiuk @XD-DENG @bdsoha Can you please review my code changes, all the tests are passed and the code is ready for review/merge ?

**Comment by mlnsharma** [comment]
Hi @potiuk @XD-DENG @bdsoha Can you please review my code changes, please let me know if anything is pending from my end?

**Comment by mlnsharma** [comment]
Hi @potiuk @jedcunningham @XD-DENG @bdsoha 
Can you please review my code changes?

**Comment by mlnsharma** [comment]
Hi @potiuk @jedcunningham @XD-DENG @bdsoha
Can you please review my code changes?

**Comment by mlnsharma** [comment]
Hi @potiuk @jedcunningham @XD-DENG @bdsoha
Can you please review my code changes?

**Comment by potiuk** [comment]
CAn you solve the conflicts please @mlnsharma before others can take a look at it?

**Pull request closed without merging**